helpers.js

```diff
@@ -0,0 +1,163 @@
+const contracts = require("truffle-workflow-compile");
+const util = require('util');
+
+const readFile = util.promisify(fs.readFile);
+const contractsCompile = util.promisify(contracts.compile);
+
+function getFormatter(style) {
+    const formatterName = style || 'stylish';
+    try {
+        return require(`eslint/lib/formatters/${formatterName}`);
+    } catch (ex) {
+        ex.message = `\nThere was a problem loading formatter option: ${style} \nError: ${
+            ex.message
+        }`;
+        throw ex;
+    }
+}
+
+function versionJSON2String(jsonResponse) {
+    return Object.keys(jsonResponse).map((key) => `${key}: ${jsonResponse[key]}`).join(', ');
+}
+
+async function printHelpMessage() {
+  return new Promise(resolve => {
+    const helpMessage = `Usage: truffle run analyze [options]
+
+Options:
+  --debug    Provide additional debug output
+  --mode { quick | full }
+             Perform quick or in-depth (full) analysis.
+  --style {stylish | unix | visualstudio | table | tap | ...},
+             Output report in the given es-lint style style.
+             See https://eslint.org/docs/user-guide/formatters/ for a full list.
+  --timeout *seconds* ,
+          Limit MythX analysis time to *s* seconds.
+          The default is 30 seconds.
+  --version show package and MythX version information
+`;
+    console.log(helpMessage);
+    resolve(null);
+  });
+}
+
+async function printVersion() {
+  const pjson = require('./package.json');
+  console.log(`${pjson.name} ${pjson.version}`);
+
+  const version = await armlet.ApiVersion();
+  
+  console.log(versionJSON2String(version))
+
+  return version;
+}
+
+function getSolidityDetails(config) {
+  const rootDir = config.working_directory;
+  const buildDir = config.contracts_build_directory;
+  // FIXME: use truffle library routine
+  const contractsDir = trufstuf.getContractsDir(rootDir);
+
+  // const expect = require("truffle-expect");
+  // FIXME: expect things
+
+  let solidityFileBase;
+  let solidityFile;
+  let buildJsonPath;
+  let buildJson;
+
+  if (config._.length === 1) {
+    buildJson = trufstuf.guessTruffleBuildJson(buildDir);
+  } else {
+    buildJson = path.basename(config._[0]);
+  }
+  solidityFileBase = path.basename(buildJson, '.json');
+
+  if (!solidityFileBase.endsWith('.sol')) {
+    solidityFileBase += '.sol';
+  }
+
+  solidityFile = path.join(contractsDir, solidityFileBase);
+  if (config.debug) {
+    config.logger.log(`Solidity file used: ${solidityFile}`);
+  }
+
+  buildJsonPath = path.join(buildDir, buildJson);
+  if (!buildJsonPath.endsWith('.json')) {
+    buildJsonPath += '.json';
+  }
+
+  return { solidityFile, buildJsonPath }; 
+}
+
+// Run Mythril Platform analyze after we have
+// ensured via compile that JSON data is there and
+// up to date.
+async function analyze(config) {
+  const { solidityFile, buildJsonPath } = getSolidityDetails(config);
+  const buildJson = await readFile(buildJsonPath);
```

Comment:
readFile is asynchronous. Synchronous one is readFileSync. In Javascript use should almost always avoid synchronous operations.
printHelpmessage and printVersion - are made async because truffle plugin expects that it will return asynchronous result. We can't mix async and sync logic

Reply:
We've just talked that truffle-analyze could analyze multiple JSON files. In this case synchronous code would block thread loop until it I/O operation would finish. This means that no other operation will be done at that moment. readFile - is asynchronous and won't block  thread loop which will allow other JSON file to start analyzing.

Reply:
I'd like to see the code side by side then for multiple JSON files. 

And this doesn't explain why we need asynchronous behavior for `printHelpmessage` and `printVersion`. Never of those is going to be run several times in parallel.

Reply:
BTW the liklihood of truffle projects have several contracts that all need to be analyzed simultaneously is pretty small. However I don't know for certain so, okay we should allow the possiblity that it can happen. But having concrete code to look at and assess would really help me understand this.

Reply:
`module.exports = async (config) => {` - is Async and it expects async results to be resolved. That's why `printHelpmessage` and `printVersion` are written in async manner. Otherwise result will never resolve. In other words `done` callback never called. `truffle run` - will exit due to timeout limit, because it expects for callback done to be called or promise to be resolved.

Reply:
Ok. Thanks for th explanation. I'd prefer then the simple-minded and simple to use promise approach which doesn't have this complexity. 

I really find it hard to believe that there is going to be significantpeformance gain (if any) going with the async approach. I think this is more about whether you have to issue `done()` or not, and  actually, I prefer the explicit and simple to understand `done()`.

Reply:
> BTW the liklihood of truffle projects have several contracts that all need to be analyzed simultaneously I suspect is pretty small. Another better approach would be to combine all of the contracts into one request and send that. This can't be done now but it is planned in the future I believe.

I don't see reason why it should not analyze multiple files. I have a case where truffle project compiles multiple smart contracts, but truffle-analyze analyzes the first smart contract it finds and does not give a damn about other smart contracts inside the project. I would say this is a bug.
To workaround this issue, I have to remove all smart contracts and run them separately in different truffle projects or one by one (deleting and pasting).

Reply:
No it is not about being able to analyze multiple files. It is about whether it is _internally_ done as one API request or several. The advantange of one API request is that MythX gets to see more of the bigger picture. And when it sees more of the bigger picture it tends to do a better job.